File: 85.maximal-rectangle.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

public class Solution85
{
    public int MaximalRectangle(char[][] matrix) {
        if(matrix == null) return 0;
        int m = matrix.Length;
        if(m == 0) return 0;
        int n = matrix[0].Length;
        if(n == 0) return 0;

        int[,] dpleft = new int[m ,n ];
        for(int i = 0; i < m; ++i)
            for(int j = 0; j < n; ++j)
            {
                if(matrix[i][j] =='0')  dpleft[i, j] = 0;
                else dpleft[i, j] = (j ==0? 0: dpleft[i, j -1]) + 1;
            }

        int result = 0;
        for(int i = 0; i < m; ++i)
            for(int j = 0; j<n; ++j)
            {
                if(matrix[i][j] == '0') 
                {
                    continue;
                }
                else
                {
                    int width = int.MaxValue;
                    int height = 0;
                    int area = 0;
                    for(int k = i; k >=0; k--)
                    {
                        height ++;
                        width = Math.Min(width, dpleft[k, j]);
                        if(width == 0) break;

                        area = Math.Max(area, height * width);
                    }
                    
                    result = Math.Max(result, area);
                }
            }

        return result;
    }
}

File: 85_maximal-rectangle.py
from typing import List


class Solution:
    def maximalRectangle(self, matrix: List[List[str]]) -> int:
        res = 0
        if not matrix: return res
        m = len(matrix)
        if m == 0:  return res
        n = len(matrix[0])
        if n == 0: return res

        dp = [[0]*(n+1) for i in range(m)]
        for i in range(m):
            for j in range(1, n+1):
                if matrix[i][j-1] == '1': 
                    dp[i][j] = dp[i][j-1] + 1

        
        for i in range(m):
            for j in range(1, n+1):
                if matrix[i][j-1] == '0':
                    continue

                width = float('inf')
                height = 0
                area = 0
                k = i
                while k >=0:
                    width = min(width, dp[k][j])
                    height +=1
                    if height == 0: 
                        break
                    area = max(area, height*width)
                    
                    k -=1

                
                res = max(res, area)

        return res
            
